compiler/rustc_mir_transform/src/pass_manager.rs

```diff
@@ -72,64 +72,73 @@ where
     }
 }
 
+/// Run the sequence of passes without validating the MIR after each pass. The MIR is still
+/// validated at the end.
+pub fn run_passes_no_validate<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &mut Body<'tcx>,
+    passes: &[&dyn MirPass<'tcx>],
+) {
+    run_passes_inner(tcx, body, passes, false);
+}
+
 pub fn run_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>, passes: &[&dyn MirPass<'tcx>]) {
+    run_passes_inner(tcx, body, passes, true);
+}
+
+fn run_passes_inner<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &mut Body<'tcx>,
+    passes: &[&dyn MirPass<'tcx>],
+    validate_each: bool,
+) {
     let start_phase = body.phase;
     let mut cnt = 0;
 
-    let validate = tcx.sess.opts.unstable_opts.validate_mir;
+    let validate = validate_each & tcx.sess.opts.unstable_opts.validate_mir;
     let overridden_passes = &tcx.sess.opts.unstable_opts.mir_enable_passes;
     trace!(?overridden_passes);
 
-    if validate {
-        validate_body(tcx, body, format!("start of phase transition from {:?}", start_phase));
```

Comment:
So we are no longer doing pre-validation to ensure we have sensible input?

Reply:
There isn't really a "pre" step - the input to all passes is the output to another. And the first pass that runs after Mir building is a nop (although I suppose it might be principled to insert an explicit invocation of the validator immediately after Mir building)

Reply:
That is relying on a non-local invariant though, so if it is ever broken it could be hard to debug. The previous check was set up to ensure that it is definitely this pass that is to blame, no matter how much things changed elsewhere in the compiler, possibly by someone who is not aware of this invariant.

Reply:
Oh hmm, had not considered that angle. Yeah, that sounds reasonable to me. Will re-add this next time I touch this logic (probably fairly soon)